docs/site/Interceptors.md

```diff
@@ -0,0 +1,328 @@
+  const result = await next();
+  console.log('convertName: after-' + invocationCtx.methodName);
+  return result;
+};
+```
+
+To leverage dependency injection, a provider class can be defined as the
+interceptor:
+
+```ts
+class NameValidator implements Provider<Interceptor> {
+  constructor(@inject('valid-names') private validNames: string[]) {}
+  value() {
+    return async <T>(
+      invocationCtx: InvocationContext,
+      next: () => ValueOrPromise<T>,
+    ) => {
+      const name = invocationCtx.args[0];
+      if (!this.validNames.includes(name)) {
+        throw new Error(
+          `Name '${name}' is not on the list of '${this.validNames}`,
+        );
+      }
+      return await next();
+    };
+  }
+}
+```
+
+## Apply interceptors
+
+Interceptors form a cascading chain of handlers around the target method
+invocation. We can apply interceptors by decorating methods/classes with
+`@intercept`.
+
+### @intercept
+
+Syntax: `@intercept(...interceptorFunctionsOrBindingKeys)`
+
+The `@intercept` decorator adds interceptors to a class or its methods including
+static and instance methods. Two flavors are accepted:
+
+- An interceptor function
+
+  ```ts
+  class MyController {
+    @intercept(log) // Use the `log` function
+    greet(name: string) {
+      return `Hello, ${name}`;
+    }
+  }
+  ```
+
+- A binding key that can be resolved to an interface function
+
+  ```ts
+  class MyController {
+    @intercept('name-validator') // Use the `name-validator` binding
+    async helloWithNameValidation(name: string) {
+      return `Hello, ${name}`;
+    }
+  }
+
+  // Bind options and provider for `NameValidator`
+  ctx.bind('valid-names').to(['John', 'Mary']);
+  ctx.bind('name-validator').toProvider(NameValidator);
+  ```
+
+### Method level interceptors
+
+A static or prototype method on a class can be decorated with `@intercept` to
+apply interceptors. For example,
+
+```ts
+class MyController {
+  @intercept(log)
+  static async greetStatic(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  static async greetStaticWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(logSync)
+  greetSync(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  greet(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept('log')
+  async hello(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  async greetWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept('log', logSync)
+  async helloWithTwoInterceptors(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  async helloWithoutInterceptors(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(changeName)
+  async helloWithChangeName(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(logWithErrorHandling)
+  async helloWithError(name: string) {
+    throw new Error('error: ' + name);
+  }
```

Comment:
I find this example too long. There are about 10 methods in an unstructured list, many of them are slight variations on the same theme. Could you please pick 3-5 most important cases and convert them into the form of a combination of short explanatory text with a code snippet?

Reply:
I did some refactoring and add more comments.

Reply:
I am afraid the changes are not enough. We are writing documentation here, the content should read as well-structured text. It's ok to use code snippets for illustrations, but these snippets should not serve as a replacement for the actual human-readable text.